Create OrderKey type, allow orderBy to take in function

leearaneta opened the pull request:
closes #300

kiaking commented on test/feature/basics/Retrieve_OrderBy.spec.js:
```diff
@@ -88,4 +100,89 @@ describe('Feature – Retrieve – Order By', () => {
+
+    const result = store.getters['entities/users/query']()
+      .with('posts')
+      .orderBy(user => user.posts.length, 'desc')
```
I think this test is false positive. OrderBy is executed before resolving relationships, so you can't count number of `posts` here. Let's remove this test for now, or could you add another test that don't use relationships?

leearaneta replied:
ahh good catch!  i removed posts from the tests completely and made this test based on `user.id`

kiaking commented on test/feature/basics/Retrieve_OrderBy.spec.js:
```diff
@@ -88,4 +100,89 @@ describe('Feature – Retrieve – Order By', () => {
+  it('can sort using a function ', () => {
+    const store = createStore([{ model: User }, { model: Post }])
+
+    store.dispatch('entities/users/create', {
```
Can we add other record, or reverse the order of the data? The original data order and the expected test result is the same, so if we have some mistake in the implementation, there might be a chance this test passes anyway.

leearaneta replied:
👍

leearaneta commented:
@kiaking thanks for the wonderful feedback!  i changed the tests up and removed `package-lock.json`.

it would be wonderful to be able to sort by nested relations - is there any way we'd be able to resolve relationships before `.orderBy()` is run?

kiaking commented:
@leearaneta 

> it would be wonderful to be able to sort by nested relations - is there any way we'd be able to resolve relationships before .orderBy() is run?

yeah this is long requested feature really. It will be possible if we could implement lazy loading feature! So it's planned to be supported in the future 👍
